snapshots/stream.go

```diff
@@ -0,0 +1,107 @@
+package snapshots
+
+import (
+	"bufio"
+	"compress/zlib"
+	"io"
+
+	protoio "github.com/gogo/protobuf/io"
+	"github.com/gogo/protobuf/proto"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+const (
+	// Do not change chunk size without new snapshot format (must be uniform across nodes)
+	snapshotChunkSize  = uint64(10e6)
+	snapshotBufferSize = int(snapshotChunkSize)
+)
+
+// StreamWriter set up a stream pipeline to serialize snapshot nodes:
+// ExportNode -> delimited Protobuf -> zlib -> buffer -> chunkWriter -> chan io.ReadCloser
+type StreamWriter struct {
+	chunkWriter *ChunkWriter
+	bufWriter   *bufio.Writer
+	zWriter     *zlib.Writer
+	protoWriter protoio.WriteCloser
+}
+
+// NewStreamWriter set up a stream pipeline to serialize snapshot nodes.
+func NewStreamWriter(ch chan<- io.ReadCloser) *StreamWriter {
+	chunkWriter := NewChunkWriter(ch, snapshotChunkSize)
+	bufWriter := bufio.NewWriterSize(chunkWriter, snapshotBufferSize)
+	zWriter, err := zlib.NewWriterLevel(bufWriter, 7)
+	if err != nil {
+		chunkWriter.CloseWithError(sdkerrors.Wrap(err, "zlib failure"))
+		return nil
+	}
+	protoWriter := protoio.NewDelimitedWriter(zWriter)
+	return &StreamWriter{
+		chunkWriter: chunkWriter,
+		bufWriter:   bufWriter,
+		zWriter:     zWriter,
+		protoWriter: protoWriter,
+	}
+}
+
+// WriteMsg implements protoio.Write interface
+func (sw *StreamWriter) WriteMsg(msg proto.Message) error {
+	return sw.protoWriter.WriteMsg(msg)
+}
+
+// Close implements io.Closer interface
+func (sw *StreamWriter) Close() error {
+	if err := sw.protoWriter.Close(); err != nil {
+		sw.chunkWriter.CloseWithError(err)
+		return err
+	}
+	if err := sw.zWriter.Close(); err != nil {
+		sw.chunkWriter.CloseWithError(err)
+		return err
+	}
+	if err := sw.bufWriter.Flush(); err != nil {
+		sw.chunkWriter.CloseWithError(err)
+		return err
+	}
+	return sw.chunkWriter.Close()
+}
+
+// CloseWithError pass error to chunkWriter
+func (sw *StreamWriter) CloseWithError(err error) {
```

Comment:
shouldn't we firstly try to run `Close` and use the error in all cases?

Reply:
I think it's fine, when an error happens we only care about closing the most underlying writer, the intermediate pipeline's partial states can be discarded. if we call all the close methods when there are multiple errors, which one do we choose?